ext/bcmath/libbcmath/src/round.c

```diff
@@ -0,0 +1,164 @@
+/* round.c: bcmath library file. */
+/*
+    Copyright (C) 1991, 1992, 1993, 1994, 1997 Free Software Foundation, Inc.
+    Copyright (C) 2000 Philip A. Nelson
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.  (LICENSE)
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to:
+
+      The Free Software Foundation, Inc.
+      59 Temple Place, Suite 330
+      Boston, MA 02111-1307 USA.
+
+    You may contact the author by:
+       e-mail:  philnelson@acm.org
+      us-mail:  Philip A. Nelson
+                Computer Science Department, 9062
+                Western Washington University
+                Bellingham, WA 98226-9062
+
+*************************************************************************/
+
+#include "bcmath.h"
+#include "private.h"
+#include <stddef.h>
+
+void bc_round(bc_num num, zend_long precision, zend_long mode, bc_num *result)
+{
+	/* clear result */
+	bc_free_num(result);
+
+	/*
+	* The following three cases result in an early return:
+	*
+	* 1. When rounding to an integer part, when trying to round to a digit
+	*    larger than the num.
+	* 2. If try to round to the finest digit of num or finer than that,
+	*    num is returned unchanged.
+	* 3. If the fraction part is all 0 from the beginning or middle to the
+	*    end, the 0's are omitted and the number of digits in num is reduced.
+	*    In that case, may end up in the same situation as 2.
+	*/
+	if (precision < 0 && num->n_len <= labs(precision)) {
```

Comment:
`labs` is wrong on Windows because a long on 64-bit Windows is still 32-bit, while a zend_long could be 64-bit.
Also, there is still undefined behaviour here if `precision == ZEND_LONG_MIN` because `-ZEND_LONG_MIN` can't be represented in a zend_long.

e.g. under `--enable-undefined-sanitizer` the following code: `echo bcround('1.01000', PHP_INT_MIN), "\n";` gives this output:
`php-src/ext/bcmath/libbcmath/src/round.c:52:37: runtime error: negation of -9223372036854775808 cannot be represented in type 'long int'; cast to an unsigned type to negate this value to itself`

I believe you should do this:
```suggestion
	if (precision < 0 && num->n_len <= (size_t) (-(precision + Z_L(1))) + 1) {
```

This gets rid of the UB error. This generates the same assembly code as this: `if (precision < 0 && num->n_len <= (size_t) -precision) {` but this does trigger UB.